### ISSUE TITLE
Strange runtime results

### ISSUE THREAD

**Issue by cszer** [csrhddlam/axial-deeplab#2]
Hello , i tested inference speed and compared it with simple torchvision resnet50 . 
I used 2080ti and pytorch 1.4
Results are : 
torchvision resnet50 - 13-15 ms
axial-resnet-s - 79-81ms
But in the paper authors show that inference speed of L model is comparable with Resnet101

**Comment by cszer** [comment]
I tested models on 224х224 torch.rand tensor

**Comment by csrhddlam** [comment]
Hello, thanks for testing it. Please note that this is a re-implementation and we haven't tried to match the inference time as the original code.

I think the main mismatch comes from implementation differences, possibly the positional encoding or batch normalization. In addition, I think the runtime won't increase much if you test it on L models -- we only change the width of the model and keep the depth the same, while ResNet doubles the depth.

**Comment by netw0rkf10w** [comment]
@csrhddlam Why don't you share the original TensorFlow implementation but a PyTorch re-implementation instead? I'm a bit confused...

**Comment by csrhddlam** [comment]
It would be almost impossible to release the original code in tensorflow, because it is Google property, and it depends on some other packages which are also Google property, e.g. stand-alone self-attention and panoptic-deeplab.

**Comment by netw0rkf10w** [comment]
@csrhddlam Hmm... Has Google recently changed their policy? They used to release TensorFlow code for their published papers...

**Comment by csrhddlam** [comment]
No, as far as I know. And sorry for the confusion. As I said, our original code depends heavily on stand-alone self-attention and panoptic-deeplab. However, they have not released their code and we are not authorized to release their code, so we cannot release our original code. Instead of waiting for their releases, we re-implement the work here in PyTorch to let the community access most details of our work as soon as possible.

**Comment by netw0rkf10w** [comment]
@csrhddlam I see. Thanks for the reply!
Good work by the way. Congratulations!

**Comment by csrhddlam** [comment]
Just investigated the inference time a bit. here is a part of my trace on a GPU.
![image](https://user-images.githubusercontent.com/16692643/91908091-ffcbd300-ec78-11ea-8b87-502bf5cb9796.png)
You can see that the relative positional embedding is taking way more time than a convolution.
![image](https://user-images.githubusercontent.com/16692643/91908341-63ee9700-ec79-11ea-9959-c85e0cf55538.png)
In addition, reshaping, squeezing, and permutating are also taking way more time than bmm, where actual computation happens.

**Comment by netw0rkf10w** [comment]
`einsum` seems to have some performance [issues](https://github.com/pytorch/pytorch/issues/32591), so maybe directly using `bmm` could be faster.

**Comment by csrhddlam** [comment]
Thanks for the pointer. I wasn't aware of the issue.

`einsum` in PyTorch looks less optimized than `einsum_v2` in TensorFlow. And I agree that directly using `bmm`, together with some smart `permute` and `view` could be faster in PyTorch.
